components/camel-dfdl/src/main/java/org/apache/camel/component/dfdl/DfdlEndpoint.java

```diff
@@ -0,0 +1,153 @@
+import org.apache.daffodil.japi.Daffodil;
+import org.apache.daffodil.japi.DataProcessor;
+import org.apache.daffodil.japi.Diagnostic;
+import org.apache.daffodil.japi.ParseResult;
+import org.apache.daffodil.japi.ProcessorFactory;
+import org.apache.daffodil.japi.UnparseResult;
+import org.apache.daffodil.japi.infoset.W3CDOMInfosetInputter;
+import org.apache.daffodil.japi.infoset.W3CDOMInfosetOutputter;
+import org.apache.daffodil.japi.io.InputSourceDataInputStream;
+
+/**
+ * Transforms fixed format data such as EDI message from/to XML using a Data Format Description Language (DFDL).
+ */
+@UriEndpoint(firstVersion = "4.11.0", scheme = "dfdl", title = "DFDL", syntax = "dfdl:schemaUri", producerOnly = true,
+             category = { Category.TRANSFORMATION })
+public class DfdlEndpoint extends ProcessorEndpoint {
+
+    @UriPath
+    @Metadata(required = true, description = "The path to the DFDL schema file.")
+    private String schemaUri;
+
+    @UriParam
+    @Metadata(defaultValue = "PARSE", description = "Transform direction. Either PARSE or UNPARSE")
+    private ParseDirection parseDirection;
+
+    @UriParam(description = "The root element of the schema to use. If not specified, the first root element in the schema will be used.",
+              defaultValue = "")
+    private String rootElement = "";
+
+    @UriParam(description = "The root namespace of the schema to use.", defaultValue = "")
+    private String rootNamespace = "";
+
+    private DataProcessor daffodilProcessor;
+
+    public DfdlEndpoint(String uri, DfdlComponent component, String schemaFile) {
+        super(uri, component);
+        this.schemaUri = schemaFile;
+    }
+
+    @Override
+    protected void onExchange(Exchange exchange) throws Exception {
+        if (getParseDirection() == ParseDirection.UNPARSE) {
+            Document xmlDocument = exchange.getIn().getBody(Document.class);
+            W3CDOMInfosetInputter inputter = new W3CDOMInfosetInputter(xmlDocument);
+            ByteArrayOutputStream bos = new ByteArrayOutputStream();
+            UnparseResult result = getDaffodilProcessor().unparse(inputter, Channels.newChannel(bos));
+            if (result.isError()) {
+                exchange.setException(new DfdlUnparseException(result));
+                return;
+            }
+            exchange.getMessage().setBody(bos);
+        } else {
+            byte[] binary = exchange.getIn().getBody(byte[].class);
+            var inputStream = new InputSourceDataInputStream(binary);
+            var outputter = new W3CDOMInfosetOutputter();
+            ParseResult result = getDaffodilProcessor().parse(inputStream, outputter);
+            if (result.isError()) {
+                exchange.setException(new DfdlParseException(result));
+                return;
+            }
+            exchange.getMessage().setBody(outputter.getResult());
+        }
+    }
+
+    public DataProcessor getDaffodilProcessor() throws IOException {
```

Comment:
This is likely better to do this in doInit so this is loaded eager during bootstrap.

Otherwise you can have a concurrency issue if multiple threads is creating this processor because they see it as null on first request.

Reply:
Done :+1: